pycbc/events/eventmgr_cython.pyx

```diff
@@ -28,3 +29,84 @@ def findchirp_cluster_over_window_cython\
             indices[j] = i
             curr_ind = i
     return j
+
+
+@boundscheck(False)
+@wraparound(False)
+@cdivision(True)
+def logsignalrateinternals_computepsignalbins(
+    double[:] pdif,
+    double[:] tdif,
+    double[:] sdif,
+    int[:] pbin,
+    int[:] tbin,
+    int[:] sbin,
+    float[:] p,
+    double[:] t,
+    float[:] s,
+    float[:] sig,
+    float[:] pref,
+    double[:] tref,
+    float[:] sref,
+    float[:] sigref,
+    double[:] shift,
+    long int[:] rtype,
+    double sense,
+    double senseref,
+    double twidth,
+    double pwidth,
+    double swidth,
+    int to_shift_ref,
+    int to_shift_ifo,
+    int length
+):
+    cdef:
+        int idx, ridx
+
+    for idx in range(length):
+        ridx = rtype[idx]
+        pdif[idx] = (pref[ridx] - p[ridx]) % (M_PI * 2)
+        tdif[idx] = shift[ridx] * to_shift_ref + tref[ridx] - shift[ridx] * to_shift_ifo - t[ridx]
+        sdif[idx] = (s[ridx] * sense * sqrt(sigref[ridx])) / (sref[ridx] * senseref * sqrt(sig[ridx]))
+
+    for idx in range(length):
+        tbin[idx] = <int>(tdif[idx] / twidth)
+        pbin[idx] = <int>(pdif[idx] / pwidth)
+        sbin[idx] = <int>(sdif[idx] / swidth)
+
+@boundscheck(False)
+@wraparound(False)
+@cdivision(True)
+def logsignalrateinternals_compute2detrate(
+    int[:] nbinned0,
+    int[:] nbinned1,
+    int[:] nbinned2,
+    long int c0_size,
+    long int c1_size,
+    long int c2_size,
+    float[:] rate,
+    long int[:] rtype,
+    float[:] sref,
+    float[:,:,::1] two_det_weights, # This declares a C-contiguous array
+    float max_penalty,
+    float ref_snr,
+    int length
+):
+    cdef:
+        int idx, ridx, id0, id1, id2
+        float rescale_fac
+
+    for idx in range(length):
+        ridx = rtype[idx]
+        id0 = nbinned0[idx] + c0_size / 2
+        id1 = nbinned1[idx] + c1_size / 2
+        id2 = nbinned2[idx] + c2_size / 2
+
+        # Long logic block.
```

Comment:
can we have an explanatory comment here? ie say what the logic ought to be implementing

Reply:
Added, thanks!